### ISSUE TITLE
Feature request: `default_copts` for a `package`

### ISSUE THREAD

**Issue by chandlerc** [bazelbuild/bazel#11280]
### Description of the problem / feature request:

`package` invocations support several `default`s but not `default_copts`. This is extremely helpful for C++ projects.

### Feature requests: what underlying problem are you trying to solve with this feature?

My C++ project has some common compile options 

### What operating system are you running Bazel on?

Linux (and many others)

### What's the output of `bazel info release`?

2020/05/02 23:49:05 Using unreleased version at commit 12d413b5fee822442980ff3dcef2fe73ed576246
development version

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

I use `bazelisk`'s `last_green` version (which you can see the git 
> Replace this line with your answer.

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

?? Not sure how this is relevant, using Bazel on a personal (non-public) project.

###  Have you found anything relevant by searching the web?

Nope.

### Any other information, logs, or outputs that you want to share?

Nope.

**Comment by jin** [comment]
I think it's unlikely that there will be more language specific hooks added into a generic API like the `package` function. Have you considered using a macro or package-local COPTS variable?

**Comment by ulfjack** [comment]
There is a `default_copts` attribute in Blaze, but not in Bazel. The reason for excluding it was to avoid having language/rule-specific code in core, so I don't think we'd want to add it back in. However, maybe we can come up with a different language-agnostic mechanism for rules to set policy at the package level? Features were supposed to be that, but there has been ~no adoption of features outside of C++ (maybe because of lack of documentation?). It might be worthwhile to look into designing something, like so:

```
package(default_config={'cc_library.copts': //foo:copts_config})
package(default_config=['//foo:copts_config'])
```

**Comment by r4nt** [comment]
Chandler is using C++ though, so he can use features.

**Comment by ulfjack** [comment]
In theory, yes. In practice, however, it requires writing a custom toolchain, and making that work across multiple repositories seems like it might be a struggle.

**Comment by chandlerc** [comment]
I'd honestly even appreciate documentation for how to use features, or how to customize my toolchain (w/o having to write a new one from scratch).

Maybe this is me missing existing docs, but reading them with fresh eyes, nothing made it obvious to me how/where I might do something like this. I'd be perfectly happy doing it if reasonably easy instead of directly munging copts.

**Comment by limdor** [comment]
We are having this discussion right now in my organization also. We want to modify copts per target base, in our case for compiler warnings. We think that the way to go are toolchain features but the main problem is as follows.

The toolchain is shared by a lot of applications inside the project, however some applications have specific requirements that need specific features.  If everyone adds features in the toolchain it gets nasty. If every teams maintains his own toolchain becomes inconsistent.
What we would need is a way to add features to a toolchain, this features would be in a different file/folder.
As far as I know it is not possible, if it is then I guess could be a solution for @chandlerc problem.

**Comment by jsharpe** [comment]
> I'd honestly even appreciate documentation for how to use features, or how to customize my toolchain (w/o having to write a new one from scratch).

Agreed, for instance I'd love to be able to add a feature for openmp to the existing builtin autodetected toolchain but I can't see how this can be done easily; looking at rules_cc's toolchain code it looks like the extension point for adding additional features in the registration process simply doesn't currently exist.

> 
> Maybe this is me missing existing docs, but reading them with fresh eyes, nothing made it obvious to me how/where I might do something like this. I'd be perfectly happy doing it if reasonably easy instead of directly munging copts.

I recently opened #11323 because features just aren't documented at all; they refer to CROSSTOOL in the current documentation which likewise isn't documented sufficiently either.

**Comment by r4nt** [comment]
Yea, making the default toolchain extensible with custom features is something that would be cool.
On the other hand I'm wondering whether we should have an extra C++ toolchain repo anyway.

**Comment by gregoryT5** [comment]
@bazelbuild/triage this issue is still relevant, please keep open

**Comment by fmeum** [comment]
@silvergasp Is https://docs.google.com/document/d/1-etGNsPneQ8W7MBMxtLloEq-Jj9ng1G-Pip-cWtTg_Y/edit something that could eventually address this by making it easy to add features to existing toolchains?

**Comment by silvergasp** [comment]
Yeah, it should certainly help, although, it differs slightly from what is proposed here. In particular, it will make it easier to reuse features. So rather than extending an existing toolchain or adding a set of default copts, it will be easier to create customized toolchains.

**Comment by Silic0nS0ldier** [comment]
`package` (`BUILD.bazel`) and `repo` (`REPO.bazel`) have `default_package_metadata` which sets the default value of `package_metadata` on targets.

Something similar to this (that is also visible to rule implementations) seems like a good primitive that `default_copts` functionality could be built off.

Something like...

```starlark
# BUILD.bazel
package(
    default_config = [":default_cc_config"],
)
cc_config(
    name = "default_cc_config",
    conlyopts = ["-std=c17"],
)
```

Then for in-scope targets, their rules would pick up all those defaults (via providers) from a new `ctx` attribute.

**Comment by limdor** [comment]
I do not think that `default_copts` is something that should be part of Bazel, it is too language specific and I believe it should go out completly at some point in time. However, I believe that the same functionality can be achieved with features. What it would make sense it to be able to define a feature for all targets in a build file (at package label).
Also for the customization that @chandlerc was talking about, I believe that https://github.com/bazelbuild/rules_cc/issues/517 would solve the issue. In this way you would be able to customize the build without having to define a full toolchain from scratch.

**Comment by Silic0nS0ldier** [comment]
Adding onto this, third party C/C++ targets are a source of significant log noise. It would be amazing if `-w` (suppress all warnings) could be scoped to all but the main module.

**Comment by fmeum** [comment]
> Adding onto this, third party C/C++ targets are a source of significant log noise. It would be amazing if `-w` (suppress all warnings) could be scoped to all but the main module.

You could set a corresponding feature globally via `--features` and then disable it in `REPO.bazel`. Perhaps the external include paths changes in Bazel 9 also help?

@keith

**Comment by keith** [comment]
Since this issue the fix for https://github.com/bazelbuild/rules_cc/issues/517 landed, so it is now much easier to inject customizable features into a toolchain without having to replace it entirely. Using that + `package(features=` is probably the best end state for this now. I also added `default_features` to gazelle so you can make sure that packages recursively get the features you want.

As far as `-w` for external repos only. Today I think the best option is still just putting `--per_file_copt=external/.*@-w --host_per_file_copt=external/.*@-w` in your `.bazelrc`, but you could also add a feature to the toolchain with the same method as above, and configure it based on the `REPO.bazel`. I also submitted https://github.com/bazelbuild/bazel/pull/28975 which intends to give you a feature automatically to base that on.

I think we should probably close this
